packages/melos/lib/src/common/utils.dart

```diff
@@ -205,6 +209,77 @@ class PromptException extends MelosException {
   String toString() => 'Was unable to prompt for input:\n$message';
 }
 
+class PersistentShell {
+  PersistentShell({
+    required MelosLogger logger,
+    String? workingDirectory,
+  }) {
+    _logger = logger;
+    _workingDirectory = workingDirectory;
+  }
+
+  final _isWindows = currentPlatform.isWindows;
+  late final MelosLogger _logger;
+  late final String? _workingDirectory;
+  late Process _process;
+  /// This list is intended to store the commands that are sent to the shell.
+  /// Currently, it is not being utilized in the code,
+  ///TODO: remove this or actually use it
+  final List<String> _commands = [];
+
+  Future<void> startShell() async {
+    final executable = _isWindows ? 'cmd.exe' : '/bin/sh';
+
+    _process = await Process.start(
+      executable,
+      _isWindows ? ['/C'] : [],
+      workingDirectory: _workingDirectory,
+    );
+
+    _process.stdout.listen(
+      (event) {
+        _logger.logWithoutNewLine(utf8.decode(event, allowMalformed: true));
+      },
+      onDone: () {
+        /// TODO: Identify and log the specific steps that have been completed
+        /// successfully.
+        _logger.success('Shell process completed some steps successfully.');
+      },
+    );
+    _process.stderr.listen(
+      (event) {
+        _logger.error(utf8.decode(event, allowMalformed: true));
+      },
+      onDone: () {
+        /// TODO: Identify and log the specific steps that have completed
+        /// with errors.
+        _logger.error('Shell process completed with errors.');
+      },
+    );
+  }
+
+  void sendCommand(String command) {
```

Comment:
We probably need to make this `Future<void>` and add a completer that we complete from the listeners (when we get the result code, more about that further down), and also we can add an assertion so that `sendCommand` isn't called when there already is one in progress.

Reply:
I tried this approach, however, it creates a lot of noise in the previous tests related to the 'multiple steps script' functionality. And it's seems safe to call `sendCommand` without the assert.
Additionally, speaking about tests, I needed to update the asset part of most of them (related to multiple steps script). This is because the way logs work in the persistent shell has changed significantly, and I'm not entirely sure how to approach this. I wasn't able to figure out how to combine the current logs with the way the persistent shell is implemented. Any suggestion? 

Any advice on how to handle this issue would be greatly appreciated.

Reply:
What noise is that approach producing? Result codes should be intercepted in the listener so that they are not printed back to the user.

For the return type of `sendCommand` isn't it fire and forget if it doesn't use the completer? And hence you can get race conditions, like queueing up a command before the previous one has given back a result code?

Reply:
I tried a new approach, and it worked out and did not impact the current tests. I think my first attempt was quite different from the one you suggested. Take a look at my last commit and let me know :)

Reply:
Yes that's pretty much how I imagined it!